src/Aevatar.CQRS/ElasticIndexingService.cs

```
@@ -125,46 +137,64 @@ public void CheckExistOrCreateStateIndex<T>(T stateBase) where T : StateBase
-            if (value is GrainId || string.Equals(propertyName, "children"))
-            {
-                document[propertyName] = JsonConvert.SerializeObject(value);
-            }
-            else
+            var stateBase = command.State;
+            var id = command.Id;
+
+            var indexName = _cqrsProvider.GetIndexName(stateBase.GetType().Name.ToLower());
+            var properties = stateBase.GetType().GetProperties();
+            var document = new Dictionary<string, object>();
+
+            foreach (var property in properties)
             {
-                if (value != null)
+                var value = property.GetValue(stateBase);
+                var propertyName = char.ToLowerInvariant(property.Name[0]) + property.Name[1..];
+
+                if (value is GrainId || string.Equals(propertyName, "children"))
                 {
-                    document.Add(propertyName, value);
+                    document[propertyName] = JsonConvert.SerializeObject(value);
+                }
+                else
+                {
+                    if (value != null)
+                    {
+                        document.Add(propertyName, value);
+                    }
                 }
             }
-        }
 
-        document.Add(CTime, DateTime.UtcNow);
+            document.Add(CTime, DateTime.UtcNow);
 
-        var response = await _elasticClient.IndexAsync(document, i => i
-            .Index(indexName)
-            .Id(id)
-        );
+            // Add a bulk index operation for this document
+            bulkDescriptor.Index<object>(op => op
+                .Index(indexName)
+                .Id(id)
+                .Document(document)
+            );
+        }
 
-        if (!response.IsValid)
+        // Execute the bulk operation
+        var bulkResponse = await _elasticClient.BulkAsync(bulkDescriptor);
+
+        // Handle response
+        if (!bulkResponse.IsValid)
```

What will be the impact to client if we silently fails the indexing? Do we need to look for logs to know that an error occurred and even worse, do we need to remedy manually when it happens? or should exception be thrown to caller (client)

---

will be handled in a separate issue #314